lms/djangoapps/course_blocks/transformers/user_partitions.py

```diff
@@ -0,0 +1,226 @@
+
+    Staff users are *not* exempted from user partition pathways.
+    """
+    VERSION = 1
+
+    @classmethod
+    def name(cls):
+        """
+        Unique identifier for the transformer's class;
+        same identifier used in setup.py.
+        """
+        return "user_partitions"
+
+    @classmethod
+    def collect(cls, block_structure):
+        """
+        Computes any information for each XBlock that's necessary to execute
+        this transformer's transform method.
+
+        Arguments:
+            block_structure (BlockStructureCollectedData)
+        """
+        # First have the split test transformer setup its group access data for each block.
+        SplitTestTransformer.collect(block_structure)
+
+        # Because user partitions are course-wide, only store data for them on the root block.
+        root_block = block_structure.get_xblock(block_structure.root_block_usage_key)
+        user_partitions = getattr(root_block, 'user_partitions', []) or []
+        block_structure.set_transformer_data(cls, 'user_partitions', user_partitions)
+
+        # If there are no user partitions, this transformation is a no-op,
+        # so there is nothing to collect.
+        if not user_partitions:
+            return
+
+        # For each block, compute merged group access. Because this is a
+        # topological sort, we know a block's parents are guaranteed to
+        # already have merged group access computed before the block itself.
+        for block_key in block_structure.topological_traversal():
+            xblock = block_structure.get_xblock(block_key)
+            parent_keys = block_structure.get_parents(block_key)
+            parent_access = [
+                block_structure.get_transformer_block_field(parent_key, cls, 'merged_group_access')
+                for parent_key in parent_keys
+            ]
+            merged_group_access = MergedGroupAccess(user_partitions, xblock, parent_access)
+            block_structure.set_transformer_block_field(block_key, cls, 'merged_group_access', merged_group_access)
+
+    def transform(self, usage_info, block_structure):
+        """
+        Mutates block_structure and block_data based on the given usage_info.
+
+        Arguments:
+            usage_info (object)
+            block_structure (BlockStructureCollectedData)
+        """
+        SplitTestTransformer().transform(usage_info, block_structure)
+
+        user_partitions = block_structure.get_transformer_data(self, 'user_partitions')
+
+        if not user_partitions:
+            return
+
+        user_groups = get_user_partition_groups(
+            usage_info.course_key, user_partitions, usage_info.user
+        )
+        block_structure.remove_block_if(
+            lambda block_key: not block_structure.get_transformer_block_field(
+                block_key, self, 'merged_group_access'
+            ).check_group_access(user_groups)
+        )
+
+
+class MergedGroupAccess(object):
```

Comment:
Is this private to this file? If so, does it get an ``_``? If not, it should probably have its own separate tests?